compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmSingleAbstractMethodLowering.kt

```diff
@@ -25,12 +32,31 @@ internal val singleAbstractMethodPhase = makeIrFilePhase(
 )
 
 private class JvmSingleAbstractMethodLowering(context: JvmBackendContext) : SingleAbstractMethodLowering(context) {
+    // The JVM BE produces inline SAM wrappers both in the scope of an inline function as well as in the scope
+    // of an inline lambda. Compare with `SamWrapperClasses.getSamWrapperClass`.
+    override val inInlineFunctionScope: Boolean
+        get() = allScopes.any { scope ->
+            scope.irElement.safeAs<IrFunction>()?.let {
+                it.isInline || it.origin == IrDeclarationOrigin.LOCAL_FUNCTION_FOR_LAMBDA
+            } ?: false
+        }
+
     override fun getWrapperVisibility(expression: IrTypeOperatorCall, scopes: List<ScopeWithIr>) =
         if (inInlineFunctionScope) DescriptorVisibilities.PUBLIC else JavaDescriptorVisibilities.PACKAGE_VISIBILITY
 
     override fun getSuperTypeForWrapper(typeOperand: IrType): IrType =
         typeOperand.erasedUpperBound.defaultType
 
+    override fun getWrappedFunctionType(klass: IrClass): IrType =
+        klass.symbol.rawType(context as JvmBackendContext)
+
+    // The constructor of a SAM wrapper is non-synthetic and should not have line numbers.
+    // Otherwise the debugger will try to step into it.
```

Comment:
I suppose we could use a test on debugger behavior, if there isn't one yet?

Reply:
There already is a test in `stepping/stepInto/samAdapter.kt`. That's how I noticed this problem in the first place. Previously, the test was passing since the constructor was synthetic (which it isn't in the JVM BE).